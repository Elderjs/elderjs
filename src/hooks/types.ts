/* eslint-disable no-use-before-define */
import { TPerfPayload, TPerfTimings } from '../utils/perf';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { THookInterface } from './hookInterface';
import {
  TRequestObject,
  SettingsOptions,
  Stack,
  TErrors,
  TServerLookupObject,
  TUserHelpers,
  PluginOptions,
} from '../utils/types';
import { RouteOptions, RoutesObject } from '../routes/types';
import { ShortcodeDefs } from '../shortcodes/types';

export type THookName =
  | 'customizeHooks'
  | 'bootstrap'
  | 'allRequests'
  | 'middleware'
  | 'request'
  | 'data'
  | 'shortcodes'
  | 'stacks'
  | 'head'
  | 'compileHtml'
  | 'html'
  | 'requestComplete'
  | 'error'
  | 'buildComplete';

export type TRunHook = (hook: THookName, params: any) => Promise<any> | any | void;

interface IHookBase {
  hook: THookName;
  name: string;
  description: string;
  priority?: Number;
}

type TVoidOrUndefined = undefined | null | void | never;
type TGenericHookReturn<T> = TVoidOrUndefined | T | Promise<TVoidOrUndefined> | Promise<T>;

type TFilteredPlugin = Omit<PluginOptions, 'init' | 'shortcodes' | 'routes' | 'hooks'>;

export interface ICustomizeHooksHook extends IHookBase {
  hook: 'customizeHooks';
  run: (params: { perf: TPerfPayload; hookInterface: THookInterface; errors: TErrors }) => TGenericHookReturn<{
    hookInterface?: THookInterface;
    errors?: TErrors;
  }>;
}
export interface IBootstrapHook extends IHookBase {
  hook: 'bootstrap';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    errors: TErrors;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    routes: RoutesObject;
    hooks: TProcessedHooksArray;
    query: any;
  }) => TGenericHookReturn<{
    errors?: TErrors;
    helpers?: TUserHelpers;
    data?: any;
    settings?: SettingsOptions;
    query?: any;
  }>;
}

export interface IAllRequestsHook extends IHookBase {
  hook: 'allRequests';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    allRequests: TRequestObject[];
    routes: RoutesObject;
    query: any;
    errors: TErrors;
  }) => TGenericHookReturn<{ errors?: TErrors; allRequests?: any }>;
}

export interface IMiddlewareHook extends IHookBase {
  hook: 'middleware';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    errors: TErrors;
    query: any;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    allRequests: TRequestObject[];
    routes: RoutesObject;
    req: any;
    next: any;
    res: any;
    serverLookupObject: TServerLookupObject;
    runHook: TRunHook;
    shortcodes: ShortcodeDefs;
    request: TRequestObject;
    router: any;
  }) => TGenericHookReturn<{
    errors?: TErrors;
    query?: any;
    helpers?: TUserHelpers;
    data?: any;
    settings?: SettingsOptions;
    allRequests?: TRequestObject[];
    routes?: RoutesObject;
    req?: any;
    next?: any;
    res?: any;
    request?: TRequestObject;
    serverLookupObject?: TServerLookupObject;
  }>;
}

export interface IRequestHook extends IHookBase {
  hook: 'request';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    request: TRequestObject;
    allRequests: TRequestObject[];
    query: any;
    errors: TErrors;
    routes: RoutesObject;
    route: RouteOptions;
  }) => TGenericHookReturn<{
    errors?: TErrors;
    helpers?: TUserHelpers;
    data?: any;
    settings?: SettingsOptions;
    request?: TRequestObject;
    route?: RouteOptions;
  }>;
}

export interface IDataHook extends IHookBase {
  hook: 'data';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    data: any;
    request: TRequestObject;
    errors: TErrors;
    helpers: TUserHelpers;
    query: any;
    routes: RoutesObject;
    cssStack: Stack;
    headStack: Stack;
    beforeHydrateStack: Stack;
    hydrateStack: Stack;
    customJsStack: Stack;
    footerStack: Stack;
    settings: SettingsOptions;
    next: any;
  }) => TGenericHookReturn<{
    errors?: TErrors;
    data?: any;
    cssStack?: Stack;
    headStack?: Stack;
    beforeHydrateStack?: Stack;
    hydrateStack?: Stack;
    customJsStack?: Stack;
    footerStack?: Stack;
  }>;
}

export interface IShortcodeHook extends IHookBase {
  hook: 'shortcodes';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    request: TRequestObject;
    query: any;
    errors: TErrors;
    cssStack: Stack;
    headStack: Stack;
    customJsStack: Stack;
    layoutHtml: any;
    shortcodes: any;
    allRequests: TRequestObject[];
  }) => TGenericHookReturn<{
    errors?: TErrors;
    layoutHtml?: any;
    cssStack?: Stack;
    headStack?: Stack;
    customJsStack?: Stack;
  }>;
}

export interface IStacksHook extends IHookBase {
  hook: 'stacks';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    request: TRequestObject;
    query: any;
    errors: TErrors;
    cssStack: Stack;
    htmlAttributesStack: Stack;
    bodyAttributesStack: Stack;
    headStack: Stack;
    beforeHydrateStack: Stack;
    hydrateStack: Stack;
    customJsStack: Stack;
    footerStack: Stack;
  }) => TGenericHookReturn<{
    errors?: TErrors;
    cssStack?: Stack;
    htmlAttributesStack?: Stack;
    bodyAttributesStack?: Stack;
    headStack?: Stack;
    beforeHydrateStack?: Stack;
    hydrateStack?: Stack;
    customJsStack?: Stack;
    footerStack?: Stack;
  }>;
}

export interface IHeadHook extends IHookBase {
  hook: 'head';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    request: TRequestObject;
    headString: string;
    query: any;
    errors: TErrors;
  }) => TGenericHookReturn<{ errors?: TErrors; headString?: string }>;
}

export interface ICompileHtmlHook extends IHookBase {
  hook: 'compileHtml';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    htmlAttributesString: string;
    bodyAttributesString: string;
    request: TRequestObject;
    headString: string;
    footerString: string;
    layoutHtml: string;
    htmlString: string;
  }) => TGenericHookReturn<{ errors?: TErrors; htmlString?: string }>;
}

export interface IHtmlHook extends IHookBase {
  hook: 'html';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    request: TRequestObject;
    htmlString: string;
    query: any;
    errors: TErrors;
  }) => TGenericHookReturn<{ errors?: TErrors; htmlString?: string }>;
}

export interface IRequestCompleteHook extends IHookBase {
  hook: 'requestComplete';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    request: TRequestObject;
    htmlString: string;
    query: any;
    settings: SettingsOptions;
    errors: TErrors;
    timings: TPerfTimings;
    data: any;
  }) => TGenericHookReturn<{ errors?: any }>;
}

export interface IErrorHook extends IHookBase {
  hook: 'error';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    request: TRequestObject;
    query: any;
    errors: TErrors;
  }) => void | undefined | Promise<void> | Promise<undefined> | any | Promise<any>;
}

export interface IBuildCompleteHook extends IHookBase {
  hook: 'buildComplete';
  run: (params: {
    plugin?: TFilteredPlugin;
    perf: TPerfPayload;
    helpers: TUserHelpers;
    data: any;
    settings: SettingsOptions;
    timings: TPerfTimings[];
    query: any;
    errors: TErrors;
    routes: RoutesObject;
    allRequests: TRequestObject[];
  }) => void | undefined | Promise<void> | Promise<undefined> | any | Promise<any>;
}

export type THooks =
  | ICustomizeHooksHook
  | IBootstrapHook
  | IAllRequestsHook
  | IMiddlewareHook
  | IRequestHook
  | IDataHook
  | IShortcodeHook
  | IStacksHook
  | IHeadHook
  | ICompileHtmlHook
  | IHtmlHook
  | IRequestCompleteHook
  | IErrorHook
  | IBuildCompleteHook;

export type HooksArray = Array<THooks>;

export type ProcessedHook<T> = T & {
  priority: Number;
  $$meta: {
    type: string;
    addedBy: string;
  };
};

export type TProcessedHook =
  | ProcessedHook<ICustomizeHooksHook>
  | ProcessedHook<IBootstrapHook>
  | ProcessedHook<IAllRequestsHook>
  | ProcessedHook<IMiddlewareHook>
  | ProcessedHook<IRequestHook>
  | ProcessedHook<IDataHook>
  | ProcessedHook<IShortcodeHook>
  | ProcessedHook<IStacksHook>
  | ProcessedHook<IHeadHook>
  | ProcessedHook<ICompileHtmlHook>
  | ProcessedHook<IHtmlHook>
  | ProcessedHook<IRequestCompleteHook>
  | ProcessedHook<IErrorHook>
  | ProcessedHook<IBuildCompleteHook>;
export type TProcessedHooksArray = Array<TProcessedHook>;
