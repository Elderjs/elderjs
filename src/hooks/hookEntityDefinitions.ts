/* istanbul ignore file */
const hookEntityDefinitions = {
  allRequests: `Every request object collected from all routes during bootstrap. It is important to note that 'allRequests' will be different at the 'request' hook during a build because the requests are split between different processes during build time using the allRequests object.`,
  hookInterface:
    'The hook interface is what defines the "contract" for each hook. It includes what properties the hook has access to and which of those properties can be mutated.',
  errors: 'An array of errors collected during the build process.',
  helpers:
    'An object of helpers loaded from `./src/helpers/index.js` in addition to the Elder.js provided helper functions.',
  data: 'An object that is passed to Svelte templates as the "data" prop.',
  settings: 'An object representing the elder.config.js and other details about the build.',
  routes: 'An object that represents all of the routes registered with Elder.js.',
  hooks: 'An array of all of the hooks that have been validated by Elder.js.',
  query: 'An object that is initially empty but is reserved for plugins and sites to add database or api access to.',
  route: 'An object representing the specific route (similar to a route.js file) for a specific request.',
  htmlAttributesStack:
    'A "stack" of attributes to be merged together that are written to the <html> tag.By default, it containt "{lang: "en"}" or an other lang set in your elder.config.js',
  bodyAttributesStack: 'A "stack" of attributes to be merged together that are written to the <body> tag.',
  headStack:
    'A "stack" of strings to be merged together (along with cssStack) that are written to the <head> tag. If you are looking to customize the head you\'re probably better looking at the "headString."',
  cssStack:
    'A "stack" of strings to be merged together to create the the cssString prop. This is mainly uses to collect the css strings emitted by SSR\'d Svelte files.',
  styleTag: 'The full <style></style> tag that is going to be written to the head of the page.',
  cssString:
    'The the css string that is wrapped in the styleTag. Added purely for convenience in case users wanted to minify the css.',
  htmlAttributesString: 'The complete html attributes as a string just before it is written.',
  bodyAttributesString: 'Body attributes as a string just before it is written.',
  headString: 'The complete <head></head> string just before it is written to the head.',
  request:
    'An object that represents the parameters required to generate a specific page on a specific route. This object originating from the all() query of a route.js file.',
  beforeHydrateStack:
    'A "stack" of generally JS script tags that are required to be loaded before a Svelte component is hydrated. This is only written to the page when a Svelte component needs to be hydrated.',
  hydrateStack: 'A "stack" Svelte components that will be hydrated.',
  customJsStack:
    'A "stack" of user specific customJs strings that will to be merged together. This is written after the Svelte components.',
  footerStack: 'A "stack" of strings to be merged together that will be added to the footer tag.',
  htmlString: 'The fully generated html for the page.',
  timings: 'An array of collected timings of the system. These are collected using the performance observer.',
  req: "The 'req' object from Express or Polka when Elder.js is being used as a server.",
  next: "The 'next' object from Express or Polka when Elder.js is being used as a server.",
  res: "The 'res' object from Express or Polka when Elder.js is being used as a server.",
  templateHtml: "The HTML string returned by the SSR'd Svelte template for the request's route.",
  shortcodes: "An array of shortcode definitions that are processed on the 'shortcodes' hook.",
  footerString: 'A HTML string that Elder.js will write to the footer.',
  layoutHtml:
    "The compiled HTML response for a route containing all of the HTML from the Route's layout and template. ",
  serverLookupObject: `A key value object where the key is the relative permalink and the object is the 'request' object used by the Elder.js server.`,
  runHook: `The function that powers hooks. 'await runhook('hookName', objectContainingProps)`,
};

// eslint-disable-next-line import/prefer-default-export
export { hookEntityDefinitions };
