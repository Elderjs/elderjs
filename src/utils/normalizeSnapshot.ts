import windowsPathFix from './windowsPathFix';

const normalizeSnapshot = (val) => {
  if (Object.prototype.toString.call(val) === '[object String]') {
    return windowsPathFix(val);
  }
  if (Object.prototype.toString.call(val) === '[object Array]') {
    return val.map(normalizeSnapshot);
  }
  if (Object.prototype.toString.call(val) === '[object Object]') {
    return Object.keys(val).reduce((out, cv) => {
      // eslint-disable-next-line no-param-reassign
      out[cv] = normalizeSnapshot(val[cv]);
      return out;
    }, {});
  }
  return val;
};

export default normalizeSnapshot;
