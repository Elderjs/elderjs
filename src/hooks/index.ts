import path from 'path';
import fs from 'fs-extra';
import { parseBuildPerf } from '../utils';
import externalHelpers from '../externalHelpers';
import { HookOptions } from './types';
import prepareShortcodeParser from '../utils/prepareShortcodeParser';
import Page from '../utils/Page';

const hooks: Array<HookOptions> = [
  {
    hook: 'bootstrap',
    name: 'elderAddExternalHelpers',
    description: 'Adds external helpers to helpers object',
    priority: 1,
    run: async ({ helpers, query, settings }) => {
      let additionalHelpers = {};
      try {
        additionalHelpers = await externalHelpers({ helpers, query, settings });
      } catch (err) {
        console.error(err);
      }
      if (additionalHelpers) {
        return {
          helpers: {
            ...helpers,
            ...additionalHelpers,
          },
        };
      }
      return null;
    },
  },
  {
    hook: 'middleware',
    name: 'elderExpressLikeMiddleware',
    description: 'An express like middleware so requests can be served by Elder.js',
    priority: 1,
    run: async ({
      serverLookupObject,
      settings,
      query,
      helpers,
      data,
      routes,
      allRequests,
      runHook,
      errors,
      shortcodes,
      req,
      next,
      res,
      request,
    }) => {
      if (req.path) {
        let reqPath = req.path;

        if (settings.server.prefix && settings.server.prefix.length > 0) {
          if (reqPath.indexOf(settings.server.prefix) !== 0) {
            return next();
          }
        }

        // see if we have a request object with the path as is. (could include / or not.)
        let requestObject = serverLookupObject[reqPath];

        if (!requestObject && reqPath[reqPath.length - 1] === '/') {
          // check the path without a slash.
          requestObject = serverLookupObject[reqPath.substring(0, reqPath.length - 1)];
        } else if (!requestObject) {
          // check the path with a slash.
          reqPath += '/';
          requestObject = serverLookupObject[reqPath];
        }

        // if we have a requestObject then we know it is for ElderGuide
        if (requestObject) {
          let route = routes[requestObject.route];
          if (request && request.route) {
            route = routes[request.route];
          }
          const forPage = {
            request: { ...requestObject, ...request },
            settings,
            query,
            helpers,
            data,
            route,
            runHook,
            allRequests,
            routes,
            errors,
            shortcodes,
          };

          const page = new Page(forPage);

          const html = await page.html();

          if (html && !res.headerSent) {
            res.setHeader('Content-Type', 'text/html');
            res.end(html);
          }
        } else {
          next();
        }
      } else {
        next();
      }
      return {};
    },
  },
  {
    hook: 'shortcodes',
    name: 'elderProcessShortcodes',
    description:
      "Builds the shortcode parser, parses shortcodes from the html returned by the route's html and appends anything needed to the stacks.",
    priority: 50,
    run: async ({
      helpers,
      data,
      settings,
      request,
      query,
      cssStack,
      headStack,
      customJsStack,
      layoutHtml,
      shortcodes,
      allRequests,
    }) => {
      const ShortcodeParser = prepareShortcodeParser({
        shortcodes,
        helpers,
        data,
        settings,
        request,
        query,
        cssStack,
        headStack,
        customJsStack,
        allRequests,
      });

      const html = await ShortcodeParser.parse(layoutHtml);

      return {
        layoutHtml: html,
        headStack,
        cssStack,
        customJsStack,
      };
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddMetaCharsetToHead',
    description: `Adds <meta charset="UTF-8" /> to the head.`,
    priority: 100,
    run: async ({ headStack }) => {
      return {
        headStack: [
          ...headStack,
          {
            source: 'elderAddMetaCharsetToHead',
            string: `<meta charset="UTF-8" />`,
            priority: 100,
          },
        ],
      };
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddMetaViewportToHead',
    description: `Adds <meta name="viewport" content="width=device-width, initial-scale=1" /> to the head.`,
    priority: 90,
    run: async ({ headStack }) => {
      return {
        headStack: [
          ...headStack,
          {
            source: 'elderAddMetaViewportToHead',
            string: `<meta name="viewport" content="width=device-width, initial-scale=1" />`,
            priority: 90,
          },
        ],
      };
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddDefaultIntersectionObserver',
    description: 'Sets up the default polyfill for the intersection observer',
    priority: 100,
    run: async ({ beforeHydrateStack }) => {
      return {
        beforeHydrateStack: [
          {
            source: 'elderAddDefaultIntersectionObserver',
            string: `<script type="text/javascript">
      if (!('IntersectionObserver' in window)) {
          var script = document.createElement("script");
          script.src = "/static/intersection-observer.js";
          document.getElementsByTagName('head')[0].appendChild(script);
      };
      </script>
      `,
            priority: 100,
          },
          ...beforeHydrateStack,
        ],
      };
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddHtmlLangAttributes',
    description: 'Add lang attributes to html according to elder.config.js',
    priority: 100,
    run: async ({ htmlAttributesStack, settings }) => {
      return {
        htmlAttributesStack: [
          ...htmlAttributesStack,
          {
            source: 'elderAddHtmlLangAttributes',
            priority: 50,
            string: `lang="${settings.lang}"`,
          },
        ],
      };
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddBodyClassAttributes',
    description: 'Add class attributes to html based on request.route',
    priority: 100,
    run: async ({ bodyAttributesStack, request }) => {
      return {
        bodyAttributesStack: [
          ...bodyAttributesStack,
          {
            source: 'elderAddBodyClassAttributes',
            priority: 50,
            string: `class="${request.route}"`,
          },
        ],
      };
    },
  },
  {
    hook: 'compileHtml',
    name: 'elderCompileHtml',
    description: 'Creates an HTML string out of the Svelte layout and stacks.',
    priority: 50,
    run: async ({ htmlAttributesString, headString, bodyAttributesString, footerString, layoutHtml }) => {
      return {
        htmlString: `<!DOCTYPE html><html ${htmlAttributesString}><head>${headString}</head><body ${bodyAttributesString}>${layoutHtml}${footerString}</body></html>`,
      };
    },
  },

  {
    hook: 'error',
    name: 'elderConsoleLogErrors',
    description: 'Log any errors to the console.',
    priority: 1,
    run: async ({ errors, request, settings }) => {
      if (!settings.worker) {
        console.error(request.permalink, errors);
      }
    },
  },
  {
    hook: 'requestComplete',
    name: 'elderWriteHtmlFileToPublic',
    description: 'Write the html output to public.',
    priority: 1,
    run: async ({ settings, request, htmlString, errors }) => {
      if (settings.build) {
        const file = path.resolve(settings.distDir, `.${request.permalink}/index.html`);
        try {
          fs.outputFileSync(file, htmlString);
        } catch (e) {
          console.log(e);
          return {
            errors: [...errors, e],
          };
        }
      }
      return null;
    },
  },
  {
    hook: 'requestComplete',
    name: 'elderDisplayRequestTime',
    description: 'Page generating timings and logging.',
    priority: 50,
    run: async ({ timings, request, settings }) => {
      if (!settings.build && process.env.NODE_ENV !== 'production') {
        console.log(`${Math.round(timings.slice(-1)[0].duration * 10) / 10}ms: \t ${request.permalink}`);
        if (settings.debug.performance) {
          const outTimings = [...timings];
          const display = outTimings.sort((a, b) => a.duration - b.duration).map((t) => ({ ...t, ms: t.duration }));

          console.table(display, ['name', 'ms']);
        }
      }
    },
  },
  {
    hook: 'buildComplete',
    name: 'elderShowParsedBuildTimes',
    description: 'A breakdown of the average times of different stages of the build.',
    priority: 50,
    run: async ({ timings, settings }) => {
      if (settings.debug.performance) {
        const buildTimings = parseBuildPerf(timings);
        console.log(buildTimings);
      }
    },
  },
  {
    hook: 'buildComplete',
    name: 'elderWriteBuildErrors',
    description: 'Writes out any errors of a build to a JSON file in the ___ELDER___ folder.',
    priority: 50,
    run: async ({ errors, settings }) => {
      if (errors && errors.length > 0) {
        const buildOutputLocation = path.resolve(settings.rootDir, `./___ELDER___/build-errors-${Date.now()}.json`);
        console.log(
          `Errors during Elder.js build. Writing details on the ${errors.length} build errors to: ${buildOutputLocation}`,
        );
        fs.writeJSONSync(buildOutputLocation, { settings, buildErrors: errors });

        errors.forEach((error) => {
          console.error(error);
          console.error(`------`);
        });
      }
    },
  },
];

export default hooks;
