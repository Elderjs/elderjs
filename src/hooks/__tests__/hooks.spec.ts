import hooks from '../index';
import normalizeSnapshot from '../../utils/normalizeSnapshot';

jest.mock('../../externalHelpers', () => () => Promise.resolve({ permalink: jest.fn() }));

jest.mock('../../utils/prepareShortcodeParser', () => ({ customJsStack, cssStack, headStack }) => ({
  parse: (templateHtml) => {
    customJsStack.push('console.log("test");');
    cssStack.push('body{display:none;}');
    headStack.push('<title>Hello</title>');
    return Promise.resolve(`..${templateHtml}..`);
  },
}));

jest.mock('../../utils/Page', () => {
  return jest.fn(() => ({
    html: () => Promise.resolve(`<html>new page</html>`),
  }));
});

jest.mock('fs-extra', () => ({
  writeJSONSync: jest.fn(),
  outputFileSync: jest
    .fn()
    .mockImplementationOnce(() => {})
    .mockImplementationOnce(() => {
      throw new Error('Failed to write');
    }),
}));

describe('#hooks', () => {
  it('has valid priority', () => {
    expect(hooks.filter((h) => h.priority < 1 && h.priority > 100)).toEqual([]);
  });
  it('matchesSnapshot', () => {
    expect(normalizeSnapshot(hooks)).toMatchSnapshot();
  });
  it('elderAddExternalHelpers', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddExternalHelpers');
    const c = await hook.run({ helpers: { old: jest.fn() }, query: {}, settings: {} });
    expect(normalizeSnapshot(c)).toMatchSnapshot();
  });
  it('elderExpressLikeMiddleware', async () => {
    const hook = hooks.find((h) => h.name === 'elderExpressLikeMiddleware');
    const router = () => 'router() was called';

    expect(await hook.run({ router })).toBe('router() was called');
  });
  it('elderProcessShortcodes', async () => {
    const hook = hooks.find((h) => h.name === 'elderProcessShortcodes');
    const headStack = [];
    const cssStack = [];
    const customJsStack = [];
    expect(
      await hook.run({
        headStack,
        cssStack,
        customJsStack,
        layoutHtml: '<html>hi</html>',
        helpers: {},
        query: {},
        settings: {},
      }),
    ).toEqual({
      cssStack: ['body{display:none;}'],
      customJsStack: ['console.log("test");'],
      headStack: ['<title>Hello</title>'],
      layoutHtml: '..<html>hi</html>..',
    });
  });
  it('elderAddMetaCharsetToHead', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddMetaCharsetToHead');
    expect(normalizeSnapshot(await hook.run({ headStack: [] }))).toMatchSnapshot();
  });
  it('elderAddMetaViewportToHead', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddMetaViewportToHead');
    expect(normalizeSnapshot(await hook.run({ headStack: [] }))).toMatchSnapshot();
  });
  it('elderCompileHtml', async () => {
    const hook = hooks.find((h) => h.name === 'elderCompileHtml');
    expect(
      await hook.run({
        request: { route: 'test' },
        headString: 'head',
        footerString: 'footer',
        layoutHtml: 'layout',
        settings: { lang: 'en' },
      }),
    ).toStrictEqual({
      htmlString: '<!DOCTYPE html><html lang="en"><head>head</head><body class="test">layoutfooter</body></html>',
    });
  });

  it('elderConsoleLogErrors', async () => {
    const hook = hooks.find((h) => h.name === 'elderConsoleLogErrors');
    expect(
      await hook.run({ settings: { worker: false }, request: { permalink: '/foo' }, errors: ['foo', 'bar'] }),
    ).toBeUndefined();
  });
  it('elderWriteHtmlFileToPublic', async () => {
    const hook = hooks.find((h) => h.name === 'elderWriteHtmlFileToPublic');
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        htmlString: '<html>string</html>',
        errors: [],
        settings: {},
      }),
    ).toBeNull();
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        htmlString: '<html>string</html>',
        errors: [],
        settings: { build: './build', locations: { public: './public' }, distDir: process.cwd() },
      }),
    ).toBeNull();
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        htmlString: '<html>string</html>',
        errors: [],
        settings: { build: './build', locations: { public: './public' }, distDir: process.cwd() },
      }),
    ).toEqual({ errors: [new Error('Failed to write')] });
  });
  it('elderDisplayRequestTime', async () => {
    const hook = hooks.find((h) => h.name === 'elderDisplayRequestTime');
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        timings: [
          { name: 'foo', duration: 500 },
          { name: 'bar', duration: 250 },
        ],
        settings: { debug: { performance: true } },
      }),
    ).toBeUndefined();
  });
  it('elderShowParsedBuildTimes', async () => {
    const hook = hooks.find((h) => h.name === 'elderShowParsedBuildTimes');
    expect(
      await hook.run({
        timings: [
          [
            { name: 'foo', duration: 500 },
            { name: 'bar', duration: 250 },
          ],
          [
            { name: 'foo', duration: 2500 },
            { name: 'bar', duration: 0 },
          ],
        ],
        settings: { debug: { performance: true } },
      }),
    ).toBeUndefined();
  });
  it('elderWriteBuildErrors', async () => {
    const hook = hooks.find((h) => h.name === 'elderWriteBuildErrors');
    expect(
      await hook.run({
        errors: ['error1', 'error2'],
        settings: { debug: { performance: true }, rootDir: process.cwd() },
      }),
    ).toBeUndefined();
  });

  describe('#elderAddCssFileToHead', () => {
    it('respects settings.css = file', async () => {
      const hook = hooks.find((h) => h.name === 'elderAddCssFileToHead');
      expect(
        await hook.run({
          errors: ['error1', 'error2'],
          headStack: [],
          settings: {
            css: 'file',
            $$internal: {
              publicCssFile: '/_elderjs/assets/svelte.123.js',
            },
          },
        }),
      ).toEqual({
        headStack: [
          {
            priority: 30,
            source: 'elderAddCssFileToHead',
            string: `<link rel="stylesheet" href="/_elderjs/assets/svelte.123.js" media="all" />`,
          },
        ],
      });
    });

    describe('#elderAddCssFileToHead', () => {
      it('respects settings.css = lazy', async () => {
        const hook = hooks.find((h) => h.name === 'elderAddCssFileToHead');
        expect(
          await hook.run({
            errors: ['error1', 'error2'],
            headStack: [],
            settings: {
              css: 'lazy',
              $$internal: {
                publicCssFile: '/_elderjs/assets/svelte.123.js',
              },
            },
          }),
        ).toEqual({
          headStack: [
            {
              priority: 30,
              source: 'elderAddCssFileToHead',
              string: `<link rel="preload" href="/_elderjs/assets/svelte.123.js" as="style" /><link rel="stylesheet" href="/_elderjs/assets/svelte.123.js" media="print" onload="this.media='all'" /><noscript><link rel="stylesheet" href="/_elderjs/assets/svelte.123.js" media="all" /></noscript>`,
            },
          ],
        });
      });
    });

    it('respects settings.css = inline', async () => {
      const hook = hooks.find((h) => h.name === 'elderAddCssFileToHead');
      expect(
        await hook.run({
          errors: ['error1', 'error2'],
          settings: { css: 'inline' },
        }),
      ).toBeUndefined();
    });
    it('respects settings.css = none', async () => {
      const hook = hooks.find((h) => h.name === 'elderAddCssFileToHead');
      expect(
        await hook.run({
          errors: ['error1', 'error2'],
          settings: { css: 'none' },
        }),
      ).toBeUndefined();
    });
  });
});
