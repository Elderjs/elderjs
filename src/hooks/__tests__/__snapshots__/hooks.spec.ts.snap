// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#hooks elderAddExternalHelpers 1`] = `
Object {
  "helpers": Object {
    "old": [MockFunction],
    "permalink": [MockFunction],
  },
}
`;

exports[`#hooks elderAddMetaCharsetToHead 1`] = `
Object {
  "headStack": Array [
    Object {
      "priority": 100,
      "source": "elderAddMetaCharsetToHead",
      "string": "<meta charset=\\"UTF-8\\" />",
    },
  ],
}
`;

exports[`#hooks elderAddMetaViewportToHead 1`] = `
Object {
  "headStack": Array [
    Object {
      "priority": 90,
      "source": "elderAddMetaViewportToHead",
      "string": "<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />",
    },
  ],
}
`;

exports[`#hooks matchesSnapshot 1`] = `
Array [
  Object {
    "description": "Adds external helpers to helpers object",
    "hook": "bootstrap",
    "name": "elderAddExternalHelpers",
    "priority": 1,
    "run": [Function],
  },
  Object {
    "description": "An express like middleware so requests can be served by Elder.js",
    "hook": "middleware",
    "name": "elderExpressLikeMiddleware",
    "priority": 1,
    "run": [Function],
  },
  Object {
    "description": "Builds the shortcode parser, parses shortcodes from the html returned by the route's html and appends anything needed to the stacks.",
    "hook": "shortcodes",
    "name": "elderProcessShortcodes",
    "priority": 50,
    "run": [Function],
  },
  Object {
    "description": "Adds <meta charset=\\"UTF-8\\" /> to the head.",
    "hook": "stacks",
    "name": "elderAddMetaCharsetToHead",
    "priority": 100,
    "run": [Function],
  },
  Object {
    "description": "Adds <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" /> to the head.",
    "hook": "stacks",
    "name": "elderAddMetaViewportToHead",
    "priority": 90,
    "run": [Function],
  },
  Object {
    "description": "Adds the css found in the svelte files to the head if 'css' in your 'elder.config.js' file is set to 'file'.",
    "hook": "stacks",
    "name": "elderAddCssFileToHead",
    "priority": 100,
    "run": [Function],
  },
  Object {
    "description": "Creates an HTML string out of the Svelte layout and stacks.",
    "hook": "compileHtml",
    "name": "elderCompileHtml",
    "priority": 50,
    "run": [Function],
  },
  Object {
    "description": "Log any errors to the console.",
    "hook": "error",
    "name": "elderConsoleLogErrors",
    "priority": 1,
    "run": [Function],
  },
  Object {
    "description": "Write the html output to public.",
    "hook": "requestComplete",
    "name": "elderWriteHtmlFileToPublic",
    "priority": 1,
    "run": [Function],
  },
  Object {
    "description": "Page generating timings and logging.",
    "hook": "requestComplete",
    "name": "elderDisplayRequestTime",
    "priority": 50,
    "run": [Function],
  },
  Object {
    "description": "A breakdown of the average times of different stages of the build.",
    "hook": "buildComplete",
    "name": "elderShowParsedBuildTimes",
    "priority": 50,
    "run": [Function],
  },
  Object {
    "description": "Writes out any errors of a build to a JSON file in the ___ELDER___ folder.",
    "hook": "buildComplete",
    "name": "elderWriteBuildErrors",
    "priority": 50,
    "run": [Function],
  },
]
`;
